Applications/Spire/Source/Ui/DropDownBox.cpp

```diff
@@ -0,0 +1,252 @@
+      set(Fill(QColor::fromRgb(0x33, 0x33, 0x33)));
+    style.get(Disabled() >> is_a<Icon>()).
+      set(Fill(QColor::fromRgb(0xC8, 0xC8, 0xC8)));
+    style.get(ReadOnly() >> is_a<Icon>()).
+      set(Visibility::NONE);
+    style.get(Any() >> (is_a<TextBox>() && !(+Any() << is_a<ListView>()))).
+      set(PaddingRight(scale_width(8)));
+    style.get(Disabled() >> (is_a<TextBox>() && !(+Any() << is_a<ListView>()))).
+      set(TextColor(QColor::fromRgb(0xC8, 0xC8, 0xC8)));
+    return style;
+  }
+}
+
+class DropDownBox::ButtonContainer : public QWidget {
+  public:
+    ButtonContainer(QWidget* parent)
+        : QWidget(parent) {
+      auto layout = new QHBoxLayout(this);
+      layout->setContentsMargins({});
+      layout->setSpacing(0);
+      m_label = make_label("");
+      m_label->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+      m_label->setAttribute(Qt::WA_TransparentForMouseEvents);
+      layout->addWidget(m_label);
+      auto icon_layout = new QVBoxLayout();
+      icon_layout->setContentsMargins({});
+      icon_layout->setSpacing(0);
+      icon_layout->addStretch();
+      auto drop_down_icon = new Icon(
+        imageFromSvg(":/Icons/dropdown-arrow.svg", scale(6, 4)));
+      drop_down_icon->setFixedSize(scale(6, 4));
+      drop_down_icon->setFocusPolicy(Qt::NoFocus);
+      icon_layout->addWidget(drop_down_icon);
+      icon_layout->addStretch();
+      layout->addLayout(icon_layout);
+    }
+
+    const std::shared_ptr<TextModel>& get_text_model() const {
+      return m_label->get_model();
+    }
+
+  private:
+    TextBox* m_label;
+};
+
+class DropDownBox::DropDownListWrapper : public QWidget {
+  public:
+    DropDownListWrapper(ListView& list_view, QWidget* parent)
+        : QWidget(parent),
+          m_list_view(&list_view) {
+      m_drop_down_list = new DropDownList(*m_list_view, parent);
+      auto list_model = m_list_view->get_list_model();
+      for(auto i = 0; i < list_model->get_size(); ++i) {
+        m_list_view->get_list_item(i)->setFocusPolicy(Qt::NoFocus);
+      }
+      m_list_view->installEventFilter(this);
+      get_panel()->installEventFilter(this);
+    }
+
+    QWidget* get_panel() const {
+      return m_drop_down_list->window();
+    }
+
+  protected:
+    bool eventFilter(QObject* watched, QEvent* event) {
+      if(watched == m_list_view) {
+        if(event->type() == QEvent::KeyPress) {
+          auto key_event = static_cast<QKeyEvent*>(event);
+          switch(key_event->key()) {
+            case Qt::Key_Tab:
+            case Qt::Key_Backtab:
+            case Qt::Key_Escape:
+              hide();
+              QCoreApplication::sendEvent(parentWidget(), event);
+              break;
+          }
+        }
+      } else if(watched == get_panel()) {
+        if(event->type() == QEvent::Close) {
+          hide();
+        } else if(event->type() == QEvent::Show) {
+          m_list_view->setFocus();
+          m_list_view->get_current_model()->set_current(
```

Comment:
What is the purpose of this? This looks like an attempt to exploit a implementation detail side-effect.